packages/auth/src/gql/getMagicAuthLinkSchema.ts

```diff
@@ -1,52 +1,67 @@
-import type { GraphQLSchemaExtension } from '@keystone-next/keystone/types';
-
+import type { ItemRootValue } from '@keystone-next/keystone/types';
+import { graphql } from '@keystone-next/keystone';
 import { AuthGqlNames, AuthTokenTypeConfig, SecretFieldImpl } from '../types';
 
 import { createAuthToken } from '../lib/createAuthToken';
 import { validateAuthToken } from '../lib/validateAuthToken';
 import { getAuthTokenErrorMessage } from '../lib/getErrorMessage';
 
+const errorCodes = ['FAILURE', 'TOKEN_EXPIRED', 'TOKEN_REDEEMED'] as const;
+
+const MagicLinkRedemptionErrorCode = graphql.enum({
+  name: 'MagicLinkRedemptionErrorCode',
+  values: graphql.enumValues(errorCodes),
+});
+
 export function getMagicAuthLinkSchema<I extends string>({
   listKey,
   identityField,
   gqlNames,
   magicAuthLink,
   magicAuthTokenSecretFieldImpl,
+  base,
 }: {
   listKey: string;
   identityField: I;
   gqlNames: AuthGqlNames;
   magicAuthLink: AuthTokenTypeConfig;
   magicAuthTokenSecretFieldImpl: SecretFieldImpl;
-}): GraphQLSchemaExtension {
+  base: graphql.BaseSchemaMeta;
+}) {
+  const RedeemItemMagicAuthTokenFailure = graphql.object<{
+    code: typeof errorCodes[number];
+    message: string;
+  }>()({
+    name: gqlNames.RedeemItemMagicAuthTokenFailure,
+    fields: {
+      code: graphql.field({ type: graphql.nonNull(MagicLinkRedemptionErrorCode) }),
+      message: graphql.field({ type: graphql.nonNull(graphql.String) }),
+    },
+  });
+  const RedeemItemMagicAuthTokenSuccess = graphql.object<{ token: string; item: ItemRootValue }>()({
+    name: gqlNames.RedeemItemMagicAuthTokenSuccess,
+    fields: {
+      token: graphql.field({ type: graphql.nonNull(graphql.String) }),
+      item: graphql.field({ type: graphql.nonNull(base.object(listKey)) }),
+    },
+  });
+  const RedeemItemMagicAuthTokenResult = graphql.union({
+    name: gqlNames.RedeemItemMagicAuthTokenResult,
+    types: [RedeemItemMagicAuthTokenSuccess, RedeemItemMagicAuthTokenFailure],
+    resolveType(val) {
+      return 'token' in val
```

Comment:
From @mitchellhamilton in a call,  this should never exist as a `null`, so this optimisation should be OK